File: Application/renderer/pipeline.cpp
#pragma once

#include "system.h"

#include <unordered_map>
using namespace std;
#include <SFML/Graphics/RenderWindow.hpp>
using namespace sf;
#include <GLM/glm.hpp>
#include <GLM/gtc/type_ptr.hpp>
using namespace glm;

#include "shader.h"
#include "framebuffer.h"
#include "camera.h"
#include "texture.h"


class ModulePipeline : public Module
{
	unordered_map<string, GLuint> Textures;

	GLuint forms, occlusion, combine, antialiasing;

	void Init()
	{
		Opengl->Init();

		Pipeline();

		Uniforms();

		Listeners();
	}

	void Update()
	{

	}

	void Listeners()
	{
		Event->Listen("WindowRecreated", [=]{
			Uniforms();
		});

		Event->Listen("WindowResize", [=]{
			Uniforms();
		});

		Event->Listen("ShaderUpdated", [=]{
			Uniforms();
		});
	}

	void Pipeline()
	{
		unordered_map<GLenum, pair<string, GLenum>> forms_targets;
		forms_targets.insert(make_pair(GL_COLOR_ATTACHMENT0, make_pair("position", GL_RGBA32F)));
		forms_targets.insert(make_pair(GL_COLOR_ATTACHMENT1, make_pair("normal",   GL_RGBA32F)));
		forms_targets.insert(make_pair(GL_COLOR_ATTACHMENT2, make_pair("albedo",   GL_RGBA32F)));
		forms_targets.insert(make_pair(GL_DEPTH_ATTACHMENT,  make_pair("depth",    GL_DEPTH_COMPONENT32F)));
		forms = CreatePass("forms.vert", "forms.frag", forms_targets);

		unordered_map<string, string> light_samplers;
		light_samplers.insert(make_pair("positions", "position"));
		light_samplers.insert(make_pair("normals",   "normal"));
		CreatePass("light.frag", "light", light_samplers);

		unordered_map<string, string> combine_samplers;
		combine_samplers.insert(make_pair("albedo", "albedo"));
		combine_samplers.insert(make_pair("lights", "light"));
		CreatePass("combine.frag", "image", combine_samplers);

		unordered_map<string, string> occlusion_samplers;
		occlusion_samplers.insert(make_pair("position_tex", "position"));
		occlusion_samplers.insert(make_pair("normal_tex",   "normal"));
		occlusion = CreatePass("occlusion.frag", "occlusion", occlusion_samplers, 0.5);
		unsigned int id = Entity->New();
		auto tex = Entity->Add<StorageTexture>(id);
		tex->Path = "noise.png";
		Entity->Get<StorageShader>(occlusion)->Samplers.insert(make_pair("noise_tex", tex->Id));

		unordered_map<string, string> blur_samplers;
		blur_samplers.insert(make_pair("image_tex",  "image"));
		blur_samplers.insert(make_pair("effect_tex", "occlusion"));
		combine = CreatePass("apply.frag", "result", blur_samplers);

		unordered_map<string, string> antialiasing_samplers;
		antialiasing_samplers.insert(make_pair("image_tex",    "result"));
		antialiasing_samplers.insert(make_pair("position_tex", "position"));
		antialiasing_samplers.insert(make_pair("normal_tex",   "normal"));
		antialiasing = CreatePass("antialiasing.frag", "antialiasing", antialiasing_samplers);

		CreatePass("screen.frag", "screen", make_pair("image_tex", "antialiasing"));
	}

	void Uniforms()
	{
		if(GLuint id = Entity->Get<StorageShader>(forms)->Program)
		{
			glUseProgram(id);
			glUniformMatrix4fv(glGetUniformLocation(id, "projection"), 1, GL_FALSE, value_ptr(Entity->Get<StorageCamera>(*Global->Get<unsigned int>("camera"))->Projection));
			glUseProgram(0);
		}

		if(GLuint id = Entity->Get<StorageShader>(antialiasing)->Program)
		{
			Vector2u Size = Global->Get<RenderWindow>("window")->getSize();
			glUseProgram(id);
			glUniform2fv(glGetUniformLocation(id, "frameBufSize"), 1, value_ptr(vec2(Size.x, Size.y)));
			glUseProgram(0);
		}

		if(GLuint id = Entity->Get<StorageShader>(combine)->Program)
		{
			Vector2u Size = Global->Get<RenderWindow>("window")->getSize();
			glUseProgram(id);
			glUniform2fv(glGetUniformLocation(id, "frameBufSize"), 1, value_ptr(vec2(Size.x, Size.y)));
			glUseProgram(0);
		}
	}

	unsigned int CreatePass(string PathFragment, string Target, pair<string, string> Sampler, float Size = 1.0)
	{
		unordered_map<string, string> samplers;
		samplers.insert(Sampler);
		return CreatePass(PathFragment, Target, samplers, Size);
	}

	unsigned int CreatePass(string PathFragment, string Target, unordered_map<string, string> Samplers = unordered_map<string, string>(), float Size = 1.0)
	{
		unordered_map<GLenum, pair<string, GLenum>> targets;
		targets.insert(make_pair(GL_COLOR_ATTACHMENT0, make_pair(Target, GL_RGBA32F)));
		return CreatePass("quad.vert", PathFragment, targets, Samplers, Size);
	}

	unsigned int CreatePass(string PathVertex, string PathFragment, unordered_map<GLenum, pair<string, GLenum>> Targets, unordered_map<string, string> Samplers = unordered_map<string, string>(), float Size = 1.0)
	{
		unsigned int id = Entity->New();
		auto shd = Entity->Add<StorageShader>(id);
		auto frb = Entity->Add<StorageFramebuffer>(id);

		shd->PathVertex   = PathVertex;
		shd->PathFragment = PathFragment;
		for(auto i : Samplers)
			shd->Samplers.insert(make_pair(i.first, Textures[i.second]));
		
		frb->Size = Size;
		for(auto i : Targets)
		{
			GLuint id = frb->AddTarget(i.first, i.second.second);
			Textures.insert(make_pair(i.second.first, id));
		}

		return id;
	}
};

File: module/renderer/pipeline.js

rendertarget("position");
rendertarget("normal");
rendertarget("albedo");
rendertarget("specular");
rendertarget("depth", "DEPTH24_STENCIL8");

renderpass("forms", {
    vertex: "forms.vert",
    fragment: "forms.frag",
    targets: {
        "COLOR_ATTACHMENT0"        : "position",
        "COLOR_ATTACHMENT1"        : "normal",
        "COLOR_ATTACHMENT2"        : "albedo",
        "COLOR_ATTACHMENT3"        : "specular",
        "DEPTH_STENCIL_ATTACHMENT" : "depth",
    },
    draw: "FORMS",
    depth: true,
    stencil: ["ALWAYS", 1, "REPLACE"],
});

renderpass("sky", {
    vertex: "forms.vert",
    fragment: "forms.frag",
    targets: {
        "COLOR_ATTACHMENT0"        : "position",
        "COLOR_ATTACHMENT1"        : "normal",
        "COLOR_ATTACHMENT2"        : "albedo",
        "DEPTH_STENCIL_ATTACHMENT" : "depth",
    },
    fallbacks: {
        "albedo": [1, 1, 1],
    },
    draw: "SKY",
    clear: false,
    stencil: ["EQUAL", 0],
});

rendertarget("light");

renderpass("lights", {
    fragment: "light.frag",
    targets: { "COLOR_ATTACHMENT0": "light" },
    samplers: { "positions": "position", "normals": "normal", "speculars": "specular" },
    fallbacks: { "light": [.5, .5, .5] },
    draw: "LIGHTS",
    stencil: ["GEQUAL", 1],
});

rendertarget("edge");

renderpass("edge", {
    fragment: "edge.frag",
    targets: { "COLOR_ATTACHMENT0": "edge" },
    samplers: { "depth_tex": "depth", "normal_tex": "normal" },
    fallbacks: { "edge": [0, 0, 0] },
});

rendertarget("image");

renderpass("combine", {
    fragment: "combine.frag",
    targets: { "COLOR_ATTACHMENT0": "image" },
    samplers: {
        "albedo": "albedo",
        "lights": "light",
        "depth": "depth",
    },
});

rendertarget("occlusion", "RGB16F", 0.75);

rendertarget("noise", "RGBA16", 0.0);
rendertargetload("noise", "noise.png", true, false, false);

renderpass("occlusion", {
    fragment: "occlusion.frag",
    targets: { "COLOR_ATTACHMENT0": "occlusion" },
    samplers: {
        "depth_tex": "depth",
        "normal_tex": "normal",
        "noise_tex": "noise",
    },
    fallbacks: { "occlusion": [1, 1, 1] },
    size: 0.75,
    stencil: ["GEQUAL", 1],
});

rendertarget("occluded");

renderpass("apply", {
    fragment: "apply.frag",
    targets: { "COLOR_ATTACHMENT0": "occluded" },
    samplers: {
        "image_tex": "image",
        "effect_tex": "occlusion",
        "noise_tex": "noise",
    },
});

rendertarget("result");

renderpass("fog", {
    fragment: "fog.frag",
    targets: { "COLOR_ATTACHMENT0": "result" },
    samplers: {
        "colors": "occluded",
        "depths": "depth",
        "lights": "light",
    },
    stencil: ["GEQUAL", 1],
});

rendertarget("temp");

renderpass("blur_u", {
    fragment: "blur_u.frag",
    targets: { "COLOR_ATTACHMENT0": "temp" },
    samplers: { "image_tex": "result" },
});

rendertarget("blur");

renderpass("blur_v", {
    fragment: "blur_v.frag",
    targets: { "COLOR_ATTACHMENT0": "blur" },
    samplers: { "image_tex": "temp" },
});

rendertarget("antialiasing");

renderpass("antialiasing", {
    fragment: "antialiasing.frag",
    targets: { "COLOR_ATTACHMENT0": "antialiasing" },
    samplers: {
        "image_tex": "result",
        "blur_tex": "blur",
        "edge_tex": "edge",
    },
    fallbacks: { "antialiasing": "result" },
});

rendertarget("tonemapped");

renderpass("tonemapping", {
    fragment: "tonemapping.frag",
    targets: { "COLOR_ATTACHMENT0": "tonemapped" },
    samplers: { "image_tex": "antialiasing" },
    fallbacks: { "tonemapped": "antialiasing" },
});

renderpass("screen", {
    fragment: "screen.frag",
    samplers: { "image_tex": "tonemapped" },
    draw: "SCREEN",
});

renderpass("preview", {
    fragment: "preview.frag",
    samplers: {
        "first": "normal",
    	"second": "depth",
    	"third": "light",
    	"fourth": "occlusion",
    },
    draw: "SCREEN",
    clear: false,
});
